ksql-engine/src/main/java/io/confluent/ksql/planner/plan/StructuredDataSourceNode.java

```diff
@@ -248,34 +249,54 @@ private KTable createKTable(
       final Serde<GenericRow> genericRowSerde,
       final Serde<GenericRow> genericRowSerdeAfterRead
   ) {
+    // to build a table we apply the following transformations:
+    // 1. Create a KStream on the changelog topic.
+    // 2. mapValues to add the ROWKEY column
+    // 3. transformValues to add the ROWTIME column. transformValues is required to access the
+    //    streams ProcessorContext which has the timestamp for the record. Also, transformValues
+    //    is only available for KStream (not KTable). This is why we have to create a KStream
+    //    first instead of a KTable.
+    // 4. mapValues to transform null records into Optional<GenericRow>.EMPTY. We eventually need
+    //    to aggregate the KStream to produce the KTable. However the KStream aggregator filters
+    //    out records with null keys or values. For tables, a null value for a key represents
+    //    that the key was deleted. So we preserve these "tombstone" records by converting them
+    //    to a not-null representation.
+    // 5. Aggregate the KStream into a KTable using a custom aggregator that handles Optional.EMPTY
     if (ksqlTable.isWindowed()) {
       return table(
           builder.stream(
               ksqlTable.getKsqlTopic().getKafkaTopicName(),
-              Consumed.with(windowedSerde, genericRowSerde).withOffsetResetPolicy(autoOffsetReset)
-          ).mapValues(windowedMapper).transformValues(new AddTimestampColumn()),
+              Consumed.with(windowedSerde, genericRowSerde).withOffsetResetPolicy(autoOffsetReset))
+              .mapValues(windowedMapper)
+              .transformValues(new AddTimestampColumn())
+              .mapValues(Optional::ofNullable),
           windowedSerde,
           genericRowSerdeAfterRead
       );
     } else {
       return table(
           builder.stream(
               ksqlTable.getKsqlTopic().getKafkaTopicName(),
-              Consumed.with(Serdes.String(), genericRowSerde).withOffsetResetPolicy(autoOffsetReset)
-          ).mapValues(nonWindowedValueMapper).transformValues(new AddTimestampColumn()),
+              Consumed.with(
+                  Serdes.String(), genericRowSerde).withOffsetResetPolicy(autoOffsetReset))
+              .mapValues(nonWindowedValueMapper)
+              .transformValues(new AddTimestampColumn())
```

Comment:
Just wondering: could we just merge these two steps (transformValues and the following mapValues) into a single transformValues? Is it just for code readability?

Reply:
The one that produces the Optional? To do that we'd either need to move the Optional.ofNullable call into one of those mappers (timestamp or key) or do the reverse. I'd rather keep Optional.ofNullable here since its a fairly subtle implementation detail of how we build tables. So it would be best to keep it next to the code thats building the ktable. And doing the reverse I think hurts readability.